module/zfs/dmu_send.c

```diff
@@ -2466,18 +2803,78 @@ receive_free(struct receive_writer_arg *rwa, struct drr_free *drrf)
 	return (err);
 }
 
+static int
+receive_object_range(struct receive_writer_arg *rwa,
+    struct drr_object_range *drror)
+{
+	int ret;
+	dmu_tx_t *tx;
+	dnode_t *mdn = NULL;
+	dmu_buf_t *db = NULL;
+	uint64_t offset;
+	boolean_t byteorder = ZFS_HOST_BYTEORDER ^
+	    !!DRR_IS_RAW_BYTESWAPPED(drror->drr_flags) ^
+	    rwa->byteswap;
+
+	/*
+	 * Since dnode block sizes are constant, we should not need to worry
+	 * about making sure that the dnode block size is the same on the
+	 * sending and receiving sides for the time being. For non-raw sends,
+	 * this does not matter (and in fact we do not send a DRR_OBJECT_RANGE
+	 * record at all). Raw sends require this record type because the
+	 * encryption parameters are used to protect an entire block of bonus
+	 * buffers. If the size of dnode blocks ever becomes variable,
+	 * handling will need to be added to ensure that dnode block sizes
+	 * match on the sending and receiving side.
+	 */
+	if (drror->drr_numslots != DNODES_PER_BLOCK ||
+	    P2PHASE(drror->drr_firstobj, DNODES_PER_BLOCK) != 0 ||
+	    (drror->drr_flags & DRR_RAW_ENCRYPTED) == 0)
+		return (SET_ERROR(EINVAL));
+
+	offset = drror->drr_firstobj * sizeof (dnode_phys_t);
+	mdn = DMU_META_DNODE(rwa->os);
+
+	tx = dmu_tx_create(rwa->os);
+	ret = dmu_tx_assign(tx, TXG_WAIT);
+	if (ret != 0) {
+		dmu_tx_abort(tx);
+		return (ret);
+	}
+
+	ret = dmu_buf_hold_by_dnode(mdn, offset, FTAG, &db,
+	    DMU_READ_PREFETCH | DMU_READ_NO_DECRYPT);
+	if (ret != 0) {
+		dmu_tx_commit(tx);
+		return (ret);
+	}
+
+	dmu_buf_will_change_crypt_params(db, tx);
+	dmu_convert_to_raw(db, byteorder, drror->drr_salt, drror->drr_iv,
+	    drror->drr_mac);
+	dmu_buf_rele(db, FTAG);
+	dmu_tx_commit(tx);
+	return (0);
+}
+
 /* used to destroy the drc_ds on error */
 static void
 dmu_recv_cleanup_ds(dmu_recv_cookie_t *drc)
 {
+	ds_hold_flags_t dsflags = (drc->drc_raw) ? 0 : DS_HOLD_FLAG_DECRYPT;
+
+	/*
+	 * Wait for our resume state to be written to disk or for us
+	 * to finish user accounting before we say we are finished
+	 * receiving the dataset.
```

Comment:
Why do we need to make sure the user accounting is on disk in the non-resuming case (when we are destroying the fs)?